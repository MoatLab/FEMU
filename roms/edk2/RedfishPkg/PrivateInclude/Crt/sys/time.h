/** @file
  Include file to support building the third-party jansson library.

 Copyright (c) 2019, Intel Corporation. All rights reserved.<BR>
 (C) Copyright 2020 Hewlett Packard Enterprise Development LP<BR>

    SPDX-License-Identifier: BSD-2-Clause-Patent

**/

#ifndef REDFISH_CRT_SYS_TIME_H_
#define REDFISH_CRT_SYS_TIME_H_

#include <Library/RedfishCrtLib.h>

#endif
