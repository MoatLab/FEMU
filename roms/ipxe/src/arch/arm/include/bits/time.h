#ifndef _BITS_TIME_H
#define _BITS_TIME_H

/** @file
 *
 * ARM-specific time API implementations
 *
 */

FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL );

#endif /* _BITS_TIME_H */
