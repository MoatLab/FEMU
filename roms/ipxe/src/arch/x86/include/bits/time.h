#ifndef _BITS_TIME_H
#define _BITS_TIME_H

/** @file
 *
 * x86-specific time API implementations
 *
 */

FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL );

#include <ipxe/rtc_time.h>

#endif /* _BITS_TIME_H */
