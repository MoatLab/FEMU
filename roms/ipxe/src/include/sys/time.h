#ifndef _SYS_TIME_H
#define _SYS_TIME_H

/** @file
 *
 * Date and time
 */

FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL );

#include <stdint.h>

/** Seconds since the Epoch
 *
 * We use a 64-bit type to avoid Y2K38 issues, since we may have to
 * handle distant future dates (e.g. X.509 certificate expiry dates).
 */
typedef int64_t time_t;

#endif /* _SYS_TIME_H */
