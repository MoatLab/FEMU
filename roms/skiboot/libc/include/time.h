/******************************************************************************
 * Copyright (c) 2004, 2008 IBM Corporation
 * All rights reserved.
 * This program and the accompanying materials
 * are made available under the terms of the BSD License
 * which accompanies this distribution, and is available at
 * http://www.opensource.org/licenses/bsd-license.php
 *
 * Contributors:
 *     IBM Corporation - initial implementation
 *****************************************************************************/
#ifndef _TIME_H
#define _TIME_H

struct tm {
	int tm_sec;
	int tm_min;
	int tm_hour;
	int tm_mday;
	int tm_mon;
	int tm_year;
};

typedef long time_t;

struct timespec {
	time_t tv_sec;        /* seconds */
	long   tv_nsec;       /* nanoseconds */
};

struct tm *gmtime_r(const time_t *timep, struct tm *result);
time_t mktime(struct tm *tm);

/* Not implemented by libc but by hosting environment, however
 * this is where the prototype is expected
 */
int nanosleep(const struct timespec *req, struct timespec *rem);

#endif /* _TIME_H */
