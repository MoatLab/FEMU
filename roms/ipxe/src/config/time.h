#ifndef CONFIG_TIME_H
#define CONFIG_TIME_H

/** @file
 *
 * Time API configuration
 *
 */

FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL );

#include <config/defaults.h>

#include <config/local/time.h>

#endif /* CONFIG_TIME_H */
