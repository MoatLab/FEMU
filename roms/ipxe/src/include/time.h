#ifndef _TIME_H
#define _TIME_H

/** @file
 *
 * Date and time
 */

FILE_LICENCE ( GPL2_OR_LATER_OR_UBDL );

#include <sys/time.h>
#include <ipxe/time.h>

/** Broken-down time */
struct tm {
	/** Seconds [0,60] */
	int tm_sec;
	/** Minutes [0,59] */
	int tm_min;
	/** Hour [0,23] */
	int tm_hour;
	/** Day of month [1,31] */
	int tm_mday;
	/** Month of year [0,11] */
	int tm_mon;
	/** Years since 1900 */
	int tm_year;
	/** Day of week [0,6] (Sunday=0) */
	int tm_wday;
	/** Day of year [0,365] */
	int tm_yday;
	/** Daylight savings flag */
	int tm_isdst;
};

/**
 * Get current time in seconds since the Epoch
 *
 * @v t			Time to fill in, or NULL
 * @ret time		Current time
 */
static inline __attribute__ (( always_inline )) time_t time ( time_t *t ) {
	time_t now;

	now = ( time_now() + time_offset );
	if ( t )
		*t = now;
	return now;
}

extern time_t mktime ( struct tm *tm );

#endif /* _TIME_H */
