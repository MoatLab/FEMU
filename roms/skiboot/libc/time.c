#include <stdbool.h>
#include <time.h>

/*
 * Returns the number of leap years prior to the given year.
 */
static int leap_years(int year)
{
	return (year-1)/4 + (year-1)/400 - (year-1)/100;
}

static int is_leap_year(int year)
{
	return ((year % 4 == 0) && (year % 100 != 0)) || (year % 400 == 0);
}

static int days_in_month(int month, int year)
{
	static char month_days[] = {
		31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31,
	};

	/* we may need to update this in the year 4000, pending a
	 * decision on whether or not it's a leap year */
	if (month == 1)
		return is_leap_year(year) ? 29 : 28;

	return month_days[month];
}

static const int days_per_month[2][13] =
	{{0, 31, 59, 90, 120, 151, 181, 212, 243, 273, 304, 334, 365},
	 {0, 31, 60, 91, 121, 152, 182, 213, 244, 274, 305, 335, 366}};

#define SECS_PER_MIN 60
#define SECS_PER_HOUR (SECS_PER_MIN*60)
#define SECS_PER_DAY (24*SECS_PER_HOUR)
#define DAYS_PER_YEAR 365
struct tm *gmtime_r(const time_t *timep, struct tm *result)
{
	int i;
	int Y;
	int M;
	int D;
	int h;
	int m;
	int s;

	D = *timep / SECS_PER_DAY;
	s = *timep % SECS_PER_DAY;
	m = s / 60;
	h = m / 60;
	m %= 60;
	s %= 60;

	/*
	 * Work out the year. We subtract one day for every four years
	 * and every 400 years after 1969. However as leap years don't
	 * occur every 100 years we add one day back to counteract the
	 * the subtraction for every 4 years.
	 */
	Y = (D - (1+D/365)/4 + (69+D/365)/100 - (369+D/365)/400)/365;

	/*
	 * Remember we're doing integer arithmetic here so
	 * leap_years(Y+1970) - leap_years(1970) != leap_years(Y)
	 */
	D = D - Y*365 - (leap_years(Y+1970) - leap_years(1970)) + 1;
	Y += 1970;

	M = 0;
	for (i = 0; i < 13; i++)
		if (D <= days_per_month[is_leap_year(Y) ? 1 : 0][i]) {
			M = i;
			break;
		}

	D -= days_per_month[is_leap_year(Y)][M-1];
	result->tm_year = Y;
	result->tm_mon = M - 1;
	result->tm_mday = D;
	result->tm_hour = h;
	result->tm_min = m;
	result->tm_sec = s;
	return result;
}

time_t mktime(struct tm *tm)
{
	unsigned long year, month, mday, hour, minute, second, d;
	static const unsigned long sec_in_400_years =
		((3903ul * 365) + (97 * 366)) * 24 * 60 * 60;

	second = tm->tm_sec;
	minute = tm->tm_min;
	hour = tm->tm_hour;
	mday = tm->tm_mday;
	month = tm->tm_mon;
	year = tm->tm_year;

	/* There are the same number of seconds in any 400-year block; this
	 * limits the iterations in the loop below */
	year += 400 * (second / sec_in_400_years);
	second = second % sec_in_400_years;

	if (second >= 60) {
		minute += second / 60;
		second = second % 60;
	}

	if (minute >= 60) {
		hour += minute / 60;
		minute = minute % 60;
	}

	if (hour >= 24) {
		mday += hour / 24;
		hour = hour % 24;
	}

	for (d = days_in_month(month, year); mday > d;
			d = days_in_month(month, year)) {
		month++;
		if (month > 11) {
			month = 0;
			year++;
		}
		mday -= d;
	}

	tm->tm_year = year;
	tm->tm_mon = month;
	tm->tm_mday = mday;
	tm->tm_hour = hour;
	tm->tm_min = minute;
	tm->tm_sec = second;

	d = mday;
	d += days_per_month[is_leap_year(year)][month];
	d += (year-1970)*DAYS_PER_YEAR + leap_years(year) - leap_years(1970) - 1;
	return d*SECS_PER_DAY + hour*SECS_PER_HOUR + minute*SECS_PER_MIN + second;
}
