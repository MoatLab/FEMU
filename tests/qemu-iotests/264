#!/usr/bin/env python3
#
# Test nbd reconnect
#
# Copyright (c) 2019 Virtuozzo International GmbH.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

import time

import iotests
from iotests import qemu_img_create, qemu_io_silent_check, file_path, \
        qemu_nbd_popen, log

iotests.script_initialize(
    supported_fmts=['qcow2'],
)

disk_a, disk_b, nbd_sock = file_path('disk_a', 'disk_b', 'nbd-sock')
nbd_uri = 'nbd+unix:///?socket=' + nbd_sock
size = 5 * 1024 * 1024
wait_limit = 3
wait_step = 0.2

qemu_img_create('-f', iotests.imgfmt, disk_a, str(size))
qemu_img_create('-f', iotests.imgfmt, disk_b, str(size))

with qemu_nbd_popen('-k', nbd_sock, '-f', iotests.imgfmt, disk_b):
    vm = iotests.VM().add_drive(disk_a)
    vm.launch()
    vm.hmp_qemu_io('drive0', 'write 0 {}'.format(size))

    vm.qmp_log('blockdev-add', filters=[iotests.filter_qmp_testfiles],
               **{'node_name': 'backup0',
                  'driver': 'raw',
                  'file': {'driver': 'nbd',
                           'server': {'type': 'unix', 'path': nbd_sock},
                           'reconnect-delay': 10}})
    vm.qmp_log('blockdev-backup', device='drive0', sync='full', target='backup0',
               speed=(1 * 1024 * 1024))

    # Wait for some progress
    t = 0
    while t < wait_limit:
        jobs = vm.qmp('query-block-jobs')['return']
        if jobs and jobs[0]['offset'] > 0:
            break
        time.sleep(wait_step)
        t += wait_step

    if jobs and jobs[0]['offset'] > 0:
        log('Backup job is started')

jobs = vm.qmp('query-block-jobs')['return']
if jobs and jobs[0]['offset'] < jobs[0]['len']:
    log('Backup job is still in progress')

vm.qmp_log('block-job-set-speed', device='drive0', speed=0)

# Emulate server down time for 1 second
time.sleep(1)

with qemu_nbd_popen('-k', nbd_sock, '-f', iotests.imgfmt, disk_b):
    e = vm.event_wait('BLOCK_JOB_COMPLETED')
    log('Backup completed: {}'.format(e['data']['offset']))
    vm.qmp_log('blockdev-del', node_name='backup0')
    vm.shutdown()
