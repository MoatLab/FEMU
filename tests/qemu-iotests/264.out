Start NBD server
{"execute": "blockdev-add", "arguments": {"driver": "raw", "file": {"driver": "nbd", "reconnect-delay": 10, "server": {"path": "TEST_DIR/PID-nbd-sock", "type": "unix"}}, "node-name": "backup0"}}
{"return": {}}
{"execute": "blockdev-backup", "arguments": {"device": "drive0", "speed": 1048576, "sync": "full", "target": "backup0"}}
{"return": {}}
Backup job is started
Kill NBD server
Backup job is still in progress
{"execute": "block-job-set-speed", "arguments": {"device": "drive0", "speed": 0}}
{"return": {}}
Start NBD server
Backup completed: 5242880
{"execute": "blockdev-del", "arguments": {"node-name": "backup0"}}
{"return": {}}
Kill NBD server
